### PR TITLE
Open shift cluster api support

### PR DESCRIPTION
**Pull request by irfanurrehman**
This PR updates the cluster-api versions to map to the 4.1.x openshift machine api.
It also updates the usage and logic to comply with the many breaking api changes that has happened across the versions.

### CODE REVIEW

**Comment by irfanurrehman** [comment]
@ading1977 @chlam4 @enlinxu This has been tested to work and needs your comments.

File: pkg/action/executor/machine_scalers.go [hunk 1]
```diff
@@ -53,69 +54,30 @@ func (client *k8sClusterApi) verifyClusterAPIEnabled() error {
...
 		return nil, err
 	}
 	for _, machine := range machineList.Items {
-		machineSpecTmp := strings.Split(*machine.Spec.ProviderID, "/")
-		if len(nodeSpecTmp) < 2 {
-			return nil, fmt.Errorf("Machine " + machine.Name + " has no valid provider ID")
-		}
-		machineProviderID := machineSpecTmp[len(machineSpecTmp)-1]
-		if machineProviderID == nodeProviderID {
+		if nodeName == machine.Status.NodeRef.Name {
```
Comment (enlinxu) [inline pkg/action/executor/machine_scalers.go]: I remembered when Michael was implementing the AWS version, the provider ID is the only attr that matches. Can we make it match both on Name or ProviderID.

File: pkg/action/executor/machine_scalers.go [hunk 2]
```diff
@@ -130,9 +131,32 @@ func (controller *machineSetController) checkPreconditions() error {
...
 // executeAction scales a MachineSet by modifying its replica count
 func (controller *machineSetController) executeAction() error {
-	desiredReplicas := controller.machineSet.Status.Replicas + controller.request.diff
+	diff := controller.request.diff
+	client := controller.request.client
+	desiredReplicas := controller.machineSet.Status.Replicas + diff
 	controller.machineSet.Spec.Replicas = &desiredReplicas
-	machineSet, err := controller.request.client.machineSet.Update(controller.machineSet)
+
+	if diff < 0 {
```
Comment (chlam4) [inline pkg/action/executor/machine_scalers.go]: Just wanted to confirm: When diff is negative, is it always -1?  Meaning scaling down by 1?  If so, then all this makes sense.

Comment (irfanurrehman) [reply pkg/action/executor/machine_scalers.go]: Yes, thats right. Its always -1.

File: pkg/action/executor/machine_scalers.go (hunk 1 above)
Comment (irfanurrehman) [reply pkg/action/executor/machine_scalers.go]: @enlinxu I don't think its a good idea. I think the `NodeRef` is an authoritative link in terms of the API. Using the providerID was anyways a hacky mechanism earlier.

**Comment by ading1977** [comment]
@irfanurrehman , I am not sure if you have answered this or not, have you tested your solution in both stitched and non-stitched environment? The host names passed to kubeturbo are different under these two circumstances, and we need to support both cases.

Thanks,
Meng

**Comment by irfanurrehman** [comment]
> @irfanurrehman , I am not sure if you have answered this or not, have you tested your solution in both stitched and non-stitched environment? The host names passed to kubeturbo are different under these two circumstances, and we need to support both cases.

@ading1977 Thanks for the additional info. Matter of fact I have not tested this under stitched env. I will figure out a mechanism to test it in a stitched env also.

**Comment by irfanurrehman** [comment]
Updated a gist for demo steps [here](https://gist.github.com/irfanurrehman/614f0250d25a026101dbb648206ca168).
Please suggest a placeholder for this.
Additionally this is tested in both stitched and non stitched env and works fine.

**Comment by irfanurrehman** [comment]
Rebased.

**Pull request merged**